### PR TITLE
Add save and sampling

### PR DESCRIPTION
**Pull request by CerebralSeed**
This imports the same save and sampling method as used in your denoising diffusion repo.

### CODE REVIEW

File: gigagan_pytorch/gigagan_pytorch.py [hunk 1]
```diff
@@ -1730,6 +1740,23 @@ def train_generator_step(
...
+    def self_save_sample(self, batch_size):
+        self.G.eval()
+        with torch.inference_mode():
+            milestone = self.steps.item() // self.save_and_sample_every
+            batches = num_to_groups(self.num_samples, batch_size)
+            assert exists(batches)
+
+            G_kwargs = dict(batch_size=batches)
+            all_images_list = list(map(lambda n: self.G(batch_size = n), batches))
+            print(all_images_list)
```
Comment (lucidrains) [inline gigagan_pytorch/gigagan_pytorch.py:1752]: stray print

**Comment by lucidrains** [comment]
@CerebralSeed nice! do you want to try doing the sampling for the upsampler too? perhaps the sampled images can have 2 columns, with original -> upsampled. original can be taken from the training dataloader for now, and extra credit would be to support a different validation dataloader specifically for eval

**Comment by lucidrains** [comment]
have you tested it? i think you should be able to see something with a toy dataset

**Comment by CerebralSeed** [comment]
> @CerebralSeed nice! do you want to try doing the sampling for the upsampler too? perhaps the sampled images can have 2 columns, with original -> upsampled. original can be taken from the training dataloader for now, and extra credit would be to support a different validation dataloader specifically for eval

I'll give it a try later this week.

**Comment by CerebralSeed** [comment]
@lucidrains Was running some tests, and, like most GANs, there are stability issues. The Generator loss keeps increasing. I noticed in their paper, they were using AdamW(see Table A2 on page 19 where they specify their settings). Or maybe Wasserstein loss might be worth trying? 

Also, was checking on the raw model sizes, as it seems the Discriminator is doing _a little too well_, relatively speaking. Sure enough, the Discriminator is almost twice the size of the Generator, while in their paper, it's about half the size or less.

```
print("Generator has", sum(p.numel() for p in gan.G.parameters()), "parameters.")
print("Discriminator has",sum(p.numel() for p in gan.D.parameters()), "parameters.")
``` 

Not sure where that's coming from, yet. 

It looks like you might have gotten the stability issues sorted here https://github.com/lucidrains/gigagan-pytorch/issues/17 So I will update and try again to see if that improves.

**Comment by lucidrains** [comment]
@CerebralSeed nice to hear you are running tests! yup, it should be a lot more stable in the latest version

**Comment by CerebralSeed** [comment]
@lucidrains  Disregard that latest commit. I copied the whole file and seem to have overwritten some updates you were making. I will push a new commit.

**Comment by CerebralSeed** [comment]
Closing this Pull request and opening a new one.
